src/main/java/com/axis/jenkins/plugins/eiffel/eiffelbroadcaster/SendEiffelEventStep.java

```diff
@@ -0,0 +1,181 @@
+
+ The above copyright notice and this permission notice shall be included in
+ all copies or substantial portions of the Software.
+
+ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ THE SOFTWARE.
+ */
+
+package com.axis.jenkins.plugins.eiffel.eiffelbroadcaster;
+
+import com.axis.jenkins.plugins.eiffel.eiffelbroadcaster.eiffel.EiffelEvent;
+import com.axis.jenkins.plugins.eiffel.eiffelbroadcaster.eiffel.EventValidationFailedException;
+import com.axis.jenkins.plugins.eiffel.eiffelbroadcaster.eiffel.InvalidJsonPayloadException;
+import com.axis.jenkins.plugins.eiffel.eiffelbroadcaster.eiffel.SchemaUnavailableException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.collect.ImmutableSet;
+import hudson.AbortException;
+import hudson.Extension;
+import hudson.model.Run;
+import hudson.model.TaskListener;
+import java.util.Map;
+import java.util.Set;
+import javax.annotation.Nonnull;
+import org.jenkinsci.plugins.workflow.steps.Step;
+import org.jenkinsci.plugins.workflow.steps.StepContext;
+import org.jenkinsci.plugins.workflow.steps.StepDescriptor;
+import org.jenkinsci.plugins.workflow.steps.StepExecution;
+import org.jenkinsci.plugins.workflow.steps.SynchronousStepExecution;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.DataBoundSetter;
+
+/**
+ * Defines a pipeline step for sending an Eiffel event expressed as a {@link Map}. By default
+ * the event passed by the user will be decorated with a CONTEXT link to the current build's
+ * {@link com.axis.jenkins.plugins.eiffel.eiffelbroadcaster.eiffel.EiffelActivityTriggeredEvent}.
+ * Optionally a CAUSE link can be created instead or the link can be omitted entirely.
+ * <pre>
+ * def event = [
+ *     "meta": [
+ *         "type": "EiffelCompositionDefinedEvent",
+ *         "version": "3.0.0",
+ *     ],
+ *     "data": [
+ *         "name": "my-composition",
+ *     ],
+ * ]
+ * def sent = sendEiffelEvent event: event, activityLinkType: "CAUSE"
+ * echo "This event was sent: ${sent}"
+ * </pre>
+ * This step returns immediately as soon as the event has been validated and put on
+ * the internal outbound queue. The actual delivery of the event to the broker might
+ * not have happened at the time of the return.
+ */
+public class SendEiffelEventStep extends Step {
+    public static final String ERROR_MESSAGE_PREFIX = "Could not publish Eiffel event";
+
+    private static final Set<String> VALID_ACTIVITY_LINK_TYPES = ImmutableSet.of("CAUSE", "CONTEXT");
```

Comment:
Yeah, this should be a custom enum type. Not sure why I rejected that option. If I ever get pipeline snippets configuration working an enum will make that much nicer too.

(Snippet generation is a bit tricky since the `event` argument is a Map expression and I don't know how to represent that in Jelly. Alternatively I could skip having a field for that argument in the Jelly and just initialize it with a dummy value, which would require some kind of custom initializer. I gave up after a while. If you have any ideas I'm listening.)

Reply:
There, now it's an EnumSet<EiffelEvent.Link.Type>.